### PR TITLE
sd-device: always translate sysname to sysfs filename

### LINKED ISSUE #23327: test-sd-device failure on arm64 / armel
Version: 251-rc2

```
 389/1046 systemd / test-sd-device                                                          FAIL              0.31s   killed by signal 6 SIGABRT
>>> PATH='/<<PKGBUILDDIR>>/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games' SYSTEMD_LANGUAGE_FALLBACK_MAP='/<<PKGBUILDDIR>>/src/locale/language-fallback-map' MALLOC_PERTURB_=73 '/<<PKGBUILDDIR>>/build-deb/test-sd-device'
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
/* test_sd_device_new_from_nulstr */
lo: setting devlink: /dev/disk/by-partuuid/1290d63a-42cc-4c71-b87c-xxxxxxxxxxxx
lo: setting devlink: /dev/disk/by-path/pci-0000:00:0f.0-scsi-0:0:0:0-part3
lo: setting devlink: /dev/disk/by-label/Arch\x20Linux
lo: setting devlink: /dev/disk/by-uuid/a07b87e5-4af5-4a59-bde9-yyyyyyyyyyyy
lo: setting devlink: /dev/disk/by-partlabel/Arch\x20Linux
lo: checking devlink: /dev/disk/by-partuuid/1290d63a-42cc-4c71-b87c-xxxxxxxxxxxx
lo: checking devlink: /dev/disk/by-path/pci-0000:00:0f.0-scsi-0:0:0:0-part3
lo: checking devlink: /dev/disk/by-label/Arch\x20Linux
lo: checking devlink: /dev/disk/by-uuid/a07b87e5-4af5-4a59-bde9-yyyyyyyyyyyy
lo: checking devlink: /dev/disk/by-partlabel/Arch\x20Linux
/* test_sd_device_enumerator_filter_subsystem */
/* test_sd_device_enumerator_subsystems */
test_sd_device_one(/sys/bus/amba)
test_sd_device_one(/sys/bus/amba/drivers)
test_sd_device_one(/sys/bus/amba/drivers/clcd-pl11x)
test_sd_device_one(/sys/bus/amba/drivers/pl061_gpio)
test_sd_device_one(/sys/bus/amba/drivers/rtc-pl031)
test_sd_device_one(/sys/bus/amba/drivers/uart-pl010)
test_sd_device_one(/sys/bus/amba/drivers/uart-pl011)
test_sd_device_one(/sys/bus/clockevents)
test_sd_device_one(/sys/bus/clocksource)
test_sd_device_one(/sys/bus/container)
test_sd_device_one(/sys/bus/cpu)
test_sd_device_one(/sys/bus/dax)
test_sd_device_one(/sys/bus/edac)
test_sd_device_one(/sys/bus/event_source)
test_sd_device_one(/sys/bus/fsl-mc)
test_sd_device_one(/sys/bus/fsl-mc/drivers)
test_sd_device_one(/sys/bus/fsl-mc/drivers/fsl_mc_allocator)
test_sd_device_one(/sys/bus/fsl-mc/drivers/fsl_mc_dprc)
test_sd_device_one(/sys/bus/genpd)
test_sd_device_one(/sys/bus/gpio)
test_sd_device_one(/sys/bus/i2c)
test_sd_device_one(/sys/bus/i2c/drivers)
test_sd_device_one(/sys/bus/i2c/drivers/dummy)
test_sd_device_one(/sys/bus/i2c/drivers/max77620)
test_sd_device_one(/sys/bus/i2c/drivers/pca953x)
test_sd_device_one(/sys/bus/i2c/drivers/rtc-ds1307)
test_sd_device_one(/sys/bus/i2c/drivers/rtc-pcf85063)
test_sd_device_one(/sys/bus/i2c/drivers/rtc-pcf8563)
test_sd_device_one(/sys/bus/i2c/drivers/simple-mfd-i2c)
test_sd_device_one(/sys/bus/mdio_bus)
test_sd_device_one(/sys/bus/mdio_bus/drivers)
test_sd_device_one(/sys/bus/mdio_bus/drivers/Generic Clause 45 PHY)
test_sd_device_one(/sys/bus/mdio_bus/drivers/Generic FE-GE Realtek PHY)
test_sd_device_one(/sys/bus/mdio_bus/drivers/Generic PHY)
test_sd_device_one(/sys/bus/mdio_bus/drivers/Qualcomm Atheros AR8030)
test_sd_device_one(/sys/bus/mdio_bus/drivers/Qualcomm Atheros AR8031!AR8033)
Assertion 'sd_device_new_from_subsystem_sysname(&dev, subsystem, name) >= 0' failed at src/libsystemd/sd-device/test-sd-device.c:76, function test_sd_device_one(). Aborting.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```
Seen on arm64 and armel
https://buildd.debian.org/status/fetch.php?pkg=systemd&arch=arm64&ver=251%7Erc2-1&stamp=1652111706&raw=0
https://buildd.debian.org/status/fetch.php?pkg=systemd&arch=armel&ver=251%7Erc2-1&stamp=1652113801&raw=0

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #23327.

### CODE REVIEW

**Review by DaanDeMeyer — APPROVED** [review]
Is there a reason we didn't do this before already?

**Review by DaanDeMeyer — APPROVED** [review]

**Comment by yuwata** [comment]
The commit message is extended. One related discussion https://github.com/systemd/systemd/pull/19566#discussion_r636141699.

File: src/libsystemd/sd-device/sd-device.c [hunk 1]
```diff
@@ -414,17 +414,21 @@ _public_ int sd_device_new_from_subsystem_sysname(
         assert_return(path_is_normalized(subsystem), -EINVAL);
         assert_return(path_is_normalized(sysname), -EINVAL);
 
+        /* translate sysname back to sysfs filename */
+        name = strdupa_safe(sysname);
```
Comment (keszybz) [inline src/libsystemd/sd-device/sd-device.c:418]: The path is limited to PATH_MAX, so the alloca should be OK here.

**Review by keszybz — APPROVED** [review]
LGTM.

**Comment by mbiebl** [comment]
Still fails:
```
 406/1041 systemd / test-sd-device                                                          FAIL              1.27s   killed by signal 6 SIGABRT
>>> PATH='/<<PKGBUILDDIR>>/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games' MALLOC_PERTURB_=61 SYSTEMD_LANGUAGE_FALLBACK_MAP='/<<PKGBUILDDIR>>/src/locale/language-fallback-map' '/<<PKGBUILDDIR>>/build-deb/test-sd-device'
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
Listing only the last 100 lines from a long log.
test_sd_device_one(/sys/module/nf_reject_ipv6)
test_sd_device_one(/sys/module/nf_tables)
test_sd_device_one(/sys/module/nfnetlink)
test_sd_device_one(/sys/module/nfnetlink_log)
test_sd_device_one(/sys/module/nft_compat)
test_sd_device_one(/sys/module/nft_counter)
test_sd_device_one(/sys/module/ofpart)
test_sd_device_one(/sys/module/omap_mailbox)
test_sd_device_one(/sys/module/onenand)
test_sd_device_one(/sys/module/orion_wdt)
test_sd_device_one(/sys/module/pcie_aspm)
test_sd_device_one(/sys/module/phylink)
test_sd_device_one(/sys/module/printk)
test_sd_device_one(/sys/module/radio_si470x_common)
test_sd_device_one(/sys/module/random)
test_sd_device_one(/sys/module/rcupdate)
test_sd_device_one(/sys/module/rcutree)
test_sd_device_one(/sys/module/sata_mv)
test_sd_device_one(/sys/module/sch_fq)
test_sd_device_one(/sys/module/scsi_mod)
test_sd_device_one(/sys/module/sd_mod)
test_sd_device_one(/sys/module/sg)
test_sd_device_one(/sys/module/slab_common)
test_sd_device_one(/sys/module/spi_nor)
test_sd_device_one(/sys/module/spi_orion)
test_sd_device_one(/sys/module/spidev)
test_sd_device_one(/sys/module/spurious)
test_sd_device_one(/sys/module/srcutree)
test_sd_device_one(/sys/module/suspend)
test_sd_device_one(/sys/module/sysrq)
test_sd_device_one(/sys/module/tcp_bbr)
test_sd_device_one(/sys/module/tcp_cubic)
test_sd_device_one(/sys/module/tda18271)
test_sd_device_one(/sys/module/tda827x)
test_sd_device_one(/sys/module/tda8290)
test_sd_device_one(/sys/module/tda9887)
test_sd_device_one(/sys/module/tea5761)
test_sd_device_one(/sys/module/tea5767)
test_sd_device_one(/sys/module/tuner_simple)
test_sd_device_one(/sys/module/tuner_xc2028)
test_sd_device_one(/sys/module/usbcore)
test_sd_device_one(/sys/module/usbserial)
test_sd_device_one(/sys/module/vt)
test_sd_device_one(/sys/module/watchdog)
test_sd_device_one(/sys/module/workqueue)
test_sd_device_one(/sys/module/x_tables)
test_sd_device_one(/sys/module/xc4000)
test_sd_device_one(/sys/module/xc5000)
test_sd_device_one(/sys/module/xen)
test_sd_device_one(/sys/module/xen_kbdfront)
test_sd_device_one(/sys/module/xt_NFLOG)
test_sd_device_one(/sys/module/xt_conntrack)
test_sd_device_one(/sys/module/xt_hashlimit)
test_sd_device_one(/sys/module/xt_multiport)
test_sd_device_one(/sys/module/xt_state)
test_sd_device_one(/sys/module/xt_tcpudp)
test_sd_device_one(/sys/module/xz_dec)
test_sd_device_one(/sys/module/zswap)
/* test_sd_device_enumerator_devices */
test_sd_device_one(/sys/devices/armv7_cortex_a9)
test_sd_device_one(/sys/devices/breakpoint)
test_sd_device_one(/sys/devices/kprobe)
test_sd_device_one(/sys/devices/platform/Fixed MDIO bus.0)
test_sd_device_one(/sys/devices/platform/Fixed MDIO bus.0/mdio_bus/fixed-0)
test_sd_device_one(/sys/devices/platform/alarmtimer)
test_sd_device_one(/sys/devices/platform/cpufreq-dt)
test_sd_device_one(/sys/devices/platform/cpuidle-armada-xp.0)
test_sd_device_one(/sys/devices/platform/pmu)
test_sd_device_one(/sys/devices/platform/reg-dummy)
test_sd_device_one(/sys/devices/platform/reg-dummy/regulator/regulator.0)
test_sd_device_one(/sys/devices/platform/serial8250)
test_sd_device_one(/sys/devices/platform/soc)
test_sd_device_one(/sys/devices/platform/soc/f1010400.devbus-bootcs)
test_sd_device_one(/sys/devices/platform/soc/f1010400.devbus-bootcs/f0000000.nor)
test_sd_device_one(/sys/devices/platform/soc/f1010600.spi)
test_sd_device_one(/sys/devices/platform/soc/f1010600.spi/spi_master/spi0)
test_sd_device_one(/sys/devices/platform/soc/f1010600.spi/spi_master/spi0/spi0.0)
test_sd_device_one(/sys/devices/platform/soc/f1010600.spi/spi_master/spi0/spi0.0/mtd/mtd0)
test_sd_device_one(/sys/devices/platform/soc/f1010600.spi/spi_master/spi0/spi0.0/mtd/mtd0ro)
test_sd_device_one(/sys/devices/platform/soc/f1100000.sa-sram0)
test_sd_device_one(/sys/devices/platform/soc/f1110000.sa-sram1)
test_sd_device_one(/sys/devices/platform/soc/f1200000.bm-bppi)
test_sd_device_one(/sys/devices/platform/soc/fff00000.bootrom)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1001400.sdramc)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1008000.l2-cache)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1010300.rtc)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1010300.rtc/rtc/rtc0)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012000.serial)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012000.serial/tty/ttyS0)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012100.serial)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012100.serial/tty/ttyS1)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012200.serial)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012200.serial/tty/ttyS2)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012300.serial)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1012300.serial/tty/ttyS3)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1018000.pin-ctrl)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1018100.gpio)
test_sd_device_one(/sys/devices/platform/soc/soc:internal-regs/f1018100.gpio/gpio/gpiochip0)
Assertion 'streq(syspath, val)' failed at src/libsystemd/sd-device/test-sd-device.c:78, function test_sd_device_one(). Aborting.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```
Full build log (with this PR applied)
https://buildd.debian.org/status/fetch.php?pkg=systemd&arch=armel&ver=251%7Erc2-2&stamp=1652283425&raw=0
https://buildd.debian.org/status/fetch.php?pkg=systemd&arch=armhf&ver=251%7Erc2-2&stamp=1652283224&raw=0

**Comment by bluca** [comment]
Slightly different assert this time, but still looks related

**Comment by yuwata** [comment]
@mbiebl Please test again.

**Comment by keszybz** [comment]
rpm-build:fedora-rawhide-aarch64 
```
/* test_sd_device_enumerator_devices */
test_sd_device_one(/sys/devices/ARMH0061:00)
test_sd_device_one(/sys/devices/ARMH0061:00/gpiochip0)
Assertion 'r == -ENOENT' failed at src/libsystemd/sd-device/test-sd-device.c:108, function test_sd_device_one(). Aborting.
â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•
```

File: src/libsystemd/sd-device/test-sd-device.c [hunk 2]
```diff
@@ -66,7 +66,9 @@ static void test_sd_device_one(sd_device *d) {
                 assert_se(r == -ENOENT);
 
         r = sd_device_get_subsystem(d, &subsystem);
-        if (r >= 0) {
+        if (r >= 0 && !streq(subsystem, "gpio")) { /* Unfortunately, there exist /sys/class/gpio and
+                                                    * /sys/bus/gpio. Hence, sd_device_new_from_subsystem_sysname()
+                                                    * and sd_device_new_from_device_id() may not work as expected. */
```
Comment (keszybz) [inline src/libsystemd/sd-device/test-sd-device.c]: Shouldn't the whole check be removed if "gpio"? We're failing because we don't get -ENOENT.

Comment (yuwata) [reply src/libsystemd/sd-device/test-sd-device.c]: Ouch. Should be fixed now.

**Comment by keszybz** [comment]
This was failing in rpm-build:fedora-rawhide-aarch64, but it's passing now, so it seems the issue is fixed.

**Comment by keszybz** [comment]
LGTM.

**Review by bluca — APPROVED** [review]

**Pull request merged**